TNT-44322 TNT-44442 Support AJO surfaces and propositions

XDex opened the pull request:
<!--- The title above will be used as a bullet point in the release notes. -->
<!--- In general, start the title with a past tense verb (i.e. added, optimized, removed.) -->
<!--- For bug fixes, start with "fixed" (i.e. fixed an issue, fixed broken.) -->
<!--- For PRs that should not be included in the release notes, attach the label "ignore-for-release" -->

## Description

<!--- Describe your changes in detail -->
Support Surfaces in personalization query requests and AJO proposition rendering
The first AJO Beta release will only support rendering of MPA page-wide propositions (SPA support will be added in subsequent releases)
Functional e2e tests and test sandbox page added in https://jira.corp.adobe.com/browse/TNT-45510

## Related Issues

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
https://jira.corp.adobe.com/browse/TNT-44322
https://jira.corp.adobe.com/browse/TNT-44442
https://jira.corp.adobe.com/browse/TNT-45303
https://jira.corp.adobe.com/browse/TNT-45510

## Motivation and Context

Support Surfaces in personalization query requests and AJO proposition rendering
The first AJO Beta release will only support rendering of MPA page-wide propositions (SPA support will be added in subsequent releases)
<!--- Why is this change required? What problem does it solve? -->

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change which does not add functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html) or I'm an Adobe employee.
- [x] I have made any necessary test changes and all tests pass.
- [x] I have run the Sandbox successfully.

jonsnyder commented on src/components/Personalization/createPersonalizationDetails.js:
```diff
@@ -20,8 +20,19 @@ import {
+import getPageLocation from "./utils/getPageLocation";
 
-export default ({ renderDecisions, decisionScopes, event, viewCache }) => {
+const buildPageSurface = window => WEB_SURFACE + getPageLocation(window);
```
Rather than passing window everywhere, you could pass this function that was already injected with window. i.e.
`components/Personalization/createBuildPageSurface.js`

```
import { WEB_SURFACE } from "./constants/surface";

export default ({ window }) => () => {
  const url = win.location.href;
  return WEB_SURFACE + url.substring(url.indexOf(":") + 3);
};
```

XDex replied:
Fixed

jonsnyder commented on src/components/Personalization/groupDecisions.js:
```diff
@@ -86,14 +88,21 @@ const isViewScope = scopeDetails =>
+// only exactly matching page surfaces supported in initial AJO Beta release
+const matchesPage = (win, surface) => surface === WEB_SURFACE + getPageLocation(win);
+
+const isMatchingSurfaceScope = (win, scope) => isSurfaceScope(scope) && matchesPage(win, scope);
```
The hash part of the page url could have changed in the time between request and processing. Would that be okay?

XDex replied:
> The hash part of the page url could have changed in the time between request and processing. Would that be okay?

@jonsnyder As the initial AJO Beta release will only support MPA surfaces, for SPA use cases content will be properly rendered only for the first SPA page matching the surface URI. There won't be any re-rendering on subsequent client-side SPA view changes.

jonsnyder reviewed (commented):
As we start supporting AJO and Target, I'd like to keep things better separated and more modular. This code will turn into a huge mess if we don't. We could even separate VEC and form-based offers. Certainly there is code that can be shared between AJO and Target, but lets try to make it so the differences are separated.

Here is one idea on how to make this more modular. Let me know what you think.

Build a chain of renderers or processors that handle the returned propositions. ie
- Redirect handler
- __view__ scope Form-based handler
- __view__ scope VEC handler
- SPA scope handler
- AJO surface handler

We could also have a set of functions that modify the request which could add the surface.

ninaceban reviewed (changes requested):
Thank you, Andrei!
Here are a few suggestions that I have:
- Surface should be part of the `scopeDetails` xdm schema, so that it gets collected in Platform
- Even if the returned proposition is coming from AJO, it should have `__view__` scope so that Alloy doesn't have to know about AJO. It will continue to auto render the page-wide scope, DOM_ACTION schema propositions that are coming from AJO (in future we could support SPA view scope )

XDex commented on src/components/DataCollector/index.js:
```diff
@@ -62,7 +63,8 @@ const createDataCollector = ({ eventManager }) => {
             renderDecisions,
-            decisionScopes
+            decisionScopes,
+            surfaces
```
@ninaceban Yes, Erik asked for us to support this from the get-go.

XDex commented on src/components/Personalization/createGetPageLocation.js:
```diff
@@ -0,0 +1,16 @@
+
+export default ({ window }) => () => {
+  const url = window.location.href;
+  return url.substring(url.indexOf(":") + 3);
```
@ninaceban This is rather an edge-case, since very few if any customers would be running their prod sites on non-standard ports, but yes, I think the port should be part of surface URI for the web channel.

XDex replied:
port is part of the surface id, as per [Surfaces pattern matching spec](https://wiki.corp.adobe.com/display/CJM/Surfaces+pattern-matching+implementation) 
Also added a test for this

XDex commented:
@ninaceban Thank you for the review!
Regarding Artur's suggestion on having a separate `surface` proposition field, and have `scope` set to `__view__` for page-wide surfaces, I have detailed it [in the wiki](https://wiki.corp.adobe.com/pages/viewpage.action?pageId=2624577900), but we still need to approve this approach with Erik and Pete. Will implement the mods, once these are green-lit.

XDex commented:
@artur-ciocanu @ninaceban @jonsnyder I've consolidated and updated the PR according to the latest requirements documented in [AJO Web SDK integration wiki](https://wiki.corp.adobe.com/display/CJM/AJO+Web+Channel+-+Web+SDK+integration#AJOWebChannelWebSDKintegration-Conclusion). Please take another look when you have some spare time. Thanks!

5amfung commented:
@XDex Is this PR in good state that @jonsnyder can approve it in the comment here?  I know this won't be merged as we'll have a separate build for alpha testing.  But it would be great to have some status to reflect that we're complete.

XDex commented:
@jonsnyder @ninaceban I have implemented the latest requested mods to sendEvent API, as per Nina's wiki and [TNT-45303](https://jira.corp.adobe.com/browse/TNT-45303), please take another look when you get a chance. Thanks!
Note: ES/EDC/AJO backend should be finally available in a staging env soon, so I'm hoping to finally be able to also add a functional test sometime this week..
cc: @5amfung @jfkhoury

carterworks commented:
What is the difference between the top-level `decisionScopes` and `personalization.decisionScopes`? Is the top-level for Target and the other for AJO? Are they identical?

ninaceban commented on sandbox/public/perf/personalization.html:
```diff
@@ -34,7 +34,9 @@
       const onSendEvent = () => {
         alloy("sendEvent", {
           renderDecisions: true,
```
I would create a separate page and leave this one as is.

XDex replied:
@ninaceban I've made changes to the sandbox and the tests to prioritize the new `personalization` config, as in the meeting we had on this change it was decided that going forward it is now the recommended way to set `decisionScopes`, and the top-level `decisionScopes` config is now a deprecated one. So, anyone looking through the code/test/sandbox should be aware of this.
@jfkhoury please let me know if my understanding is wrong and we should revert all the mods relating to this.

ninaceban replied:
In the current PR we should only add the `personalization` config to support surfaces, in a following PR we will address the deprecation strategy. Even though we are recommending that the new way of querying personalized content is by sending the `decisionScopes` in the `personalization` config. We still need to be able to validate that the existing functionality works properly. That is why it is so important that we leave the sandbox app and functional tests as are, and add new page/functional tests to test the new behavior.
cc: @jfkhoury

jfkhoury replied:
This comment addresses all the `Sandbox` related comments:

We need to tackled to goals here:

1. Test the new `personalization` option.
2. Confirm that the `decisionsScopes` option still works, until we decide to deprecate it. (By the way we need a ticket to do so in v 3.0)

So I suggest we do the following:

1. I agree with @ninaceban, let's leave these pages alone for now; maybe add an inline comment that this `decisionScopes` option will be deprecated in v 3.0.
2. Let's add a brand new AJO Surfaces/Web Channel Sandbox page that showcases the surfaces support and use the new `personalization` option.
3. I would like to see a test/sandbox page/something that showcases the usage of `decisionsScopes` and `personalization.decisionsScopes` and see what happens.

Thank you @XDex !!

XDex replied:
@jfkhoury It's probably a misunderstanding on my end of the decisions taken in the meeting we had on nesting `decisionScopes` in `personalization` - I thought we'll be deprecating the top-level `decisionScopes` in my PR and not a separate one. Let me revert the sandbox mods and add inline comments to the API as you suggest.
Regarding p.2 - I've already added two e2e functional tests for AJO and a dedicated page to the sandbox, please take a look.
On p.3 - I've also previously added a test where we're checking that the top-level `decisionScopes` option is working as expected, either standalone or together with the new `personalization.decisionScopes`

ninaceban commented on sandbox/src/PersonalizationFormBased.js:
```diff
@@ -13,7 +13,10 @@ const metadata = {
 
 const usePropositions = () => {
   const [propositions, setPropositions] = useState(undefined);
```
Same here, I would have a dedicated page for the new `personalization -> surface` functionality. We could deprecate them in a separate PR.

XDex replied:
As I mentioned in the previous comment reply, we do need to also modify all existing sandbox pages to prioritize the new `personalization -> decisionScopes` format. However, as part of AJO tests, I will also be adding a dedicated sandbox page for AJO

XDex replied:
Reverted

ninaceban commented on sandbox/src/personalizationAnalyticsClientSideHelper.js:
```diff
@@ -64,14 +65,17 @@ const sendEvent = ({
 
 export const getFormBasedOffer = () => {
   sendEvent({
     eventType: "form-based-offer",
```
Same here, let's leave everything as is.

XDex replied:
Reverted

ninaceban commented on sandbox/src/useSendPageViewEvent.js:
```diff
@@ -19,7 +19,9 @@ export default ({
   viewName,
   data = {},
   xdm = {},
```
Same here.

XDex replied:
Reverted

ninaceban commented on src/components/DataCollector/validateUserEventOptions.js:
```diff
@@ -27,6 +27,10 @@ export default ({ options }) => {
     data: objectOf({}),
     renderDecisions: boolean(),
     decisionScopes: arrayOf(string()).uniqueItems(),
```
I am wondering if we shouldn't make `decisionScopes` and `personalization` mutually exclusive

XDex replied:
I don't think we should do that, since we want to maintain backwards compatibility at least initially. If we do decide to enforce mutual exclusivity, it should be done as part of a future task for deprecating the top-level `decisionScopes` option.

ninaceban commented on src/components/Personalization/createPersonalizationDetails.js:
```diff
@@ -10,7 +10,8 @@ OF ANY KIND, either express or implied. See the License for the specific languag
 */
 
-import { includes } from "../../utils";
+import { includes, isNonEmptyString, isNonEmptyArray } from "../../utils";
```
Any reason we can't have `buildPageSurface, normalizeSurfaces` directly in `./utils`? You can still have a separate file, but it is easier to control everything through a top level file/module.

XDex replied:
@ninaceban Yes, in a previous comment @jonsnyder requested that we have all surface-related utils encapsulated in a separate module.

ninaceban commented on src/components/Personalization/groupDecisions.js:
```diff
@@ -11,6 +11,7 @@ governing permissions and limitations under the License.
 */
 
 import { isNonEmptyArray, includes } from "../../utils";
```
Same here, any reason we can't have `isPageWideSurface` in `./utils`.

XDex replied:
Yes, see previous comment.

ninaceban commented on src/components/Personalization/groupDecisions.js:
```diff
@@ -86,14 +87,17 @@ const isViewScope = scopeDetails =>
   scopeDetails.characteristics.scopeType &&
   scopeDetails.characteristics.scopeType === VIEW_SCOPE_TYPE;
 
```
We could move this to `./utils`

XDex replied:
Yes, but Jon previously suggested we keep all surface-related utils in a single module

ninaceban commented on src/components/Personalization/utils/surfaceUtils.js:
```diff
@@ -0,0 +1,113 @@
+};
+
+const parseSurface = surfaceString => {
+  const matched = surfaceString.match(SURFACE_REGEX);
```
This looks really, really, really bad. Could we use `if/else` and maybe extract some code to helper functions. Otherwise this code is pretty cryptic.

ninaceban replied:
Also instead of calling `toLowerCase()` you could probably, make sure the initial value is lower cased.

XDex replied:
Sure, I've extracted helper functions
Regarding `toLowerCase()` - I can't make the whole initial value lower case, since path and fragment should still remain case-sensitive

ninaceban replied:
is it the fragment case-sensitive at authoring mode too?

XDex replied:
Fragments are not yet supported in the UI, but they are expected to be case-sensitive, yes

ninaceban commented on src/components/Personalization/utils/surfaceUtils.js:
```diff
@@ -0,0 +1,113 @@
+  !!scope &&
+  scope.indexOf(WEB + SURFACE_TYPE_DELIMITER) === 0 &&
+  scope.indexOf(FRAGMENT_DELIMITER) === -1;
+
```
I am not following, why we need this dependency injection here. Can't we send just the `surfaces` and page location and return a complex object like {surfaces, validation errors}. The caller of this function can decide to log or do something else. Otherwise `normalizeSurfaces` looks like an utility, but does a lot of stuff including logging. Testing all this via mocking etc, should be pretty painful.

XDex replied:
This is so we're not leaking any surface normalization-related errors to the caller. Do we really need to handle surface related valdation error logging in `createPersonalizationDetails.createQueryDetails`? At that level, we're only interested in getting valid and normalized surfaces and shouldn't be handling any validation-specific errors.
Testing is also not a problem, as we can just mock `logger.warn` as I've done in the tests

ninaceban commented on test/functional/specs/Personalization/C28756.js:
```diff
@@ -32,7 +32,9 @@ test("Test C28756: A form based offer should return if event command contains it
   const alloy = createAlloyProxy();
   await alloy.configure(config);
   const result = await alloy.sendEvent({
```
Revert, let's not touch existing tests in this PR.

XDex replied:
Reverted

ninaceban commented on test/functional/specs/Personalization/C44363.js:
```diff
@@ -27,7 +27,9 @@ test("Test C44363: Return proposition when QA mode set up with token of experien
 
   const result = await alloy.sendEvent({
     renderDecisions: true,
```
Same here.

XDex replied:
Reverted

ninaceban commented on test/functional/specs/Personalization/C44363.js:
```diff
@@ -46,7 +48,9 @@ test("Test C44363: Return proposition  when QA mode set up with token of experie
   });
   const result = await alloy.sendEvent({
     renderDecisions: true,
```
And here.

XDex replied:
Done

ninaceban commented on test/functional/specs/Personalization/C5805676.js:
```diff
@@ -54,7 +54,9 @@ test("Test C5805676: Merged metric propositions should be delivered", async () =
   const alloy = createAlloyProxy();
   await alloy.configure(config);
   const eventResult = await alloy.sendEvent({
```
And here

XDex replied:
Done

ninaceban commented on test/functional/specs/Personalization/C6364799.js:
```diff
@@ -52,7 +52,9 @@ test("Test C6364799: applyPropositions should render html-content-item schemas w
   const alloy = createAlloyProxy();
   await alloy.configure(config);
   const result = await alloy.sendEvent({
```
And here

XDex replied:
Done

ninaceban commented on test/functional/specs/Personalization/C782718.js:
```diff
@@ -146,7 +146,6 @@ const simulateViewChange = async (alloy, personalizationPayload) => {
   // sendEvent at a view change, this shouldn't request any target data, it should use the existing cache
   const resultingObject = await alloy.sendEvent({
     renderDecisions: true,
```
Please revert.

XDex replied:
Done

ninaceban commented on test/functional/specs/Personalization/C782719.js:
```diff
@@ -88,7 +88,6 @@ const simulateViewChange = async alloy => {
   // sendEvent at a view change, this shouldn't request any target data, it should use the existing cache
   const resultingObject = await alloy.sendEvent({
     renderDecisions: false,
```
Please revert

XDex replied:
Done

ninaceban commented on test/unit/specs/components/DataCollector/index.spec.js:
```diff
@@ -60,7 +60,8 @@ describe("Event Command", () => {
       expect(event.setUserData).toHaveBeenCalledWith(data);
       expect(eventManager.sendEvent).toHaveBeenCalledWith(event, {
         renderDecisions: true,
```
Reevert

XDex replied:
I have reverted the changes related to the new option from the API and the sandbox, but as @jfkhoury mentioned in a previous comment, we need these to be present in tests, in order to confirm that both the top-level `decisionScopes` and the new `personalization.decisionScopes` options are working as expected.

ninaceban commented on test/unit/specs/components/DataCollector/index.spec.js:
```diff
@@ -69,13 +70,39 @@ describe("Event Command", () => {
   it("sends event with decisionScopes parameter when decisionScopes is not empty", () => {
     const options = {
       renderDecisions: true,
```
Revert

ninaceban commented on test/unit/specs/components/DataCollector/index.spec.js:
```diff
@@ -69,13 +70,39 @@ describe("Event Command", () => {
 
     return sendEventCommand.run(options).then(result => {
       expect(eventManager.sendEvent).toHaveBeenCalledWith(event, {
         renderDecisions: true,
```
Revert

ninaceban commented on test/unit/specs/components/DataCollector/index.spec.js:
```diff
@@ -91,7 +118,8 @@ describe("Event Command", () => {
     return sendEventCommand.run({}).then(() => {
       expect(eventManager.sendEvent).toHaveBeenCalledWith(event, {
         renderDecisions: false,
```
Revert

ninaceban commented on test/unit/specs/components/Personalization/createApplyPropositions.spec.js:
```diff
@@ -53,13 +53,13 @@ describe("Personalization::createApplyPropositions", () => {
     executeDecisions.and.returnValue(
       Promise.resolve(PAGE_WIDE_SCOPE_DECISIONS)
     );
```
Please changes in this file. It just adds noise. If you want to fix it add a separate PR.

XDex replied:
@ninaceban It's just a minor fix, can we please leave it?

ninaceban commented on test/unit/specs/components/Personalization/createAutoRenderingHandler.spec.js:
```diff
@@ -12,10 +12,14 @@ governing permissions and limitations under the License.
   SCOPES_FOO1_FOO2_DECISIONS
 } from "./responsesMock/eventResponses";
 import createAutoRenderingHandler from "../../../../../src/components/Personalization/createAutoRenderingHandler";
+import { isPageWideSurface } from "../../../../../src/components/Personalization/utils/surfaceUtils";
```
As mentioned ensure that everything is imported form `./utils`

XDex replied:
See previous comment on this

ninaceban commented on test/unit/specs/components/Personalization/createAutoRenderingHandler.spec.js:
```diff
@@ -12,10 +12,14 @@ governing permissions and limitations under the License.
 } from "./responsesMock/eventResponses";
 import createAutoRenderingHandler from "../../../../../src/components/Personalization/createAutoRenderingHandler";
+import { isPageWideSurface } from "../../../../../src/components/Personalization/utils/surfaceUtils";
+
```
Why we have this function redeclared here, when a similar is present in prod code.

XDex replied:
I figured it doesn't make sense to export internal `isPageWideScope` from `groupDecisions` if it's just used in a test and not in actual code.

ninaceban replied:
add it in `Personalization/utils` and reuse it in both places.

XDex replied:
Fixed

ninaceban commented on test/unit/specs/components/Personalization/utils/surfaceUtils.spec.js:
```diff
@@ -0,0 +1,256 @@
+    expect(isPageWideSurface("webapp://domain.com#view")).toBeFalse();
+  });
+
+  it("expands fragment surfaces", () => {
```
As mentioned earlier `normalizeSurfaces` could be much simpler if we do not use dep injection.

XDex replied:
Replied in the earlier comment

ninaceban reviewed (changes requested):
@XDex I have checked your PR. While functionally it does the right thing there are a lot of changes that are not relevant to adding `personalization -> surfaces`. Let's keep the PR focused.

There are a lot of files and hard to follow, what is relevant and what is additional refactoring. I like the idea of improving code, but let's have separate PRs focused on that.

Also let's cover with tests the functionality you have added. Since adding `personalization -> surfaces` and `personalization -> decisionScopes` is backward compatible, most of the existing test should work without any changes. You could add a separate functional test with `personalization -> surfaces & decisionScopes` just to prove that this combination also works.

ninaceban commented on src/components/Personalization/utils/surfaceUtils.js:
```diff
@@ -0,0 +1,113 @@
+    : null;
+};
+
+const stringifySurface = surface =>
```
@XDex in this repo we favor string templates, even if other `stackoverflow` users might not share this view.

XDex replied:
@ninaceban Hmm, I see that in `utils/injectStorage` for example, string concatenation is used.. But let me replace it with string templates as you suggest.

ninaceban commented on src/components/Personalization/utils/surfaceUtils.js:
```diff
@@ -0,0 +1,118 @@
+  isNonEmptyString(surfaceTypeMatch) ? surfaceTypeMatch.toLowerCase() : "";
+const getAuthority = authorityMatch =>
+  isNonEmptyString(authorityMatch) ? authorityMatch.toLowerCase() : "";
+const getPath = pathMatch =>
```
and here as well

XDex replied:
Fixed

XDex commented on src/components/Personalization/utils/surfaceUtils.js:
```diff
@@ -0,0 +1,118 @@
+
+const getSurfaceType = surfaceTypeMatch =>
+  isNonEmptyString(surfaceTypeMatch) ? surfaceTypeMatch.toLowerCase() : "";
+const getAuthority = authorityMatch =>
```
Fixed

XDex commented:
@ninaceban @jonsnyder @jfkhoury I've addressed the review comments and rebased on latest `main`. Please take another look when you get a chance. Thank you!

XDex commented on src/components/DataCollector/index.js:
```diff
@@ -30,7 +30,7 @@ const createDataCollector = ({ eventManager }) => {
             mergeId,
             renderDecisions = false,
             decisionScopes = [],
```
Done

jonsnyder commented:
Closing in favor of #920
